### PR TITLE
Add multiplexing support and parallelize document validation

### PR DESCRIPTION
**Pull request by MohamedSabthar**
## Purpose

Fixes: [#3569](https://github.com/ballerina-platform/ballerina-standard-library/issues/3569)
Fixes: [#3430](https://github.com/ballerina-platform/ballerina-standard-library/issues/3430)

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

File: ballerina/modules/parser/parser.bal [hunk 1]
```diff
@@ -16,25 +16,31 @@
 
 public class Parser {
     private Lexer lexer;
-    private DocumentNode document;
+    private map<FragmentNode> fragments;
+    private map<OperationNode> operations;
+    private ErrorDetail[] errors;
+    private boolean isFirstAnonymousOperation;
```
Comment (MohamedSabthar) [inline ballerina/modules/parser/parser.bal:22]: Can initialise this inline

File: ballerina/engine.bal [hunk 2]
```diff
@@ -114,7 +129,67 @@ isolated class Engine {
...
             }
         }
-        return validationErrors.length() == 0 ? () : getOutputObjectFromErrorDetail(validationErrors);
+
+        final readonly & int? maxQueryDepth = self.maxQueryDepth.cloneReadOnly();
+        final readonly & __Schema schema = self.schema;
+        final readonly & map<json>? vars = variables.cloneReadOnly();
+        final readonly & boolean introspection = self.introspection;
+
+        worker AA returns ErrorDetail[] {
```
Comment (MohamedSabthar) [inline ballerina/engine.bal:138]: rename worker names

File: ballerina/engine.bal [hunk 3]
```diff
@@ -114,7 +129,67 @@ isolated class Engine {
...
+            var validator = new QueryDepthValidatorVisitor(maxQueryDepth, nodeModifierContext);
+            document.accept(validator);
+            return validator.getErrors() ?: [];
+        }
+
+        worker B returns ErrorDetail[] {
+            var validator = new SubscriptionValidatorVisitor(nodeModifierContext);
```
Comment (MohamedSabthar) [inline ballerina/engine.bal:145]: Could remove var here

File: ballerina/websocket_utils.bal [hunk 4]
```diff
@@ -20,32 +20,59 @@ import graphql.parser;
...
-    if sourceStream is stream<any, error?> {
-        record {|any value;|}|error? next = sourceStream.next();
-        while next !is () {
-            any|error resultValue = next is error ? next : next.value;
-            OutputObject outputObject = engine.getResult(node, context, resultValue);
-            if outputObject.hasKey(DATA_FIELD) || outputObject.hasKey(ERRORS_FIELD) {
-                check sendWebSocketResponse(caller, customHeaders, WS_NEXT, outputObject.toJson(), connectionId);
+                                   parser:OperationNode node, SubscriptionHandler? subscriptionHandler)
+returns websocket:Error? {
+    worker A returns websocket:Error? {
```
Comment (MohamedSabthar) [inline ballerina/websocket_utils.bal:25]: move it to function and call start & handle errors

File: ballerina/modules/parser/parser.bal (hunk 1 above)
Comment (MohamedSabthar) [reply ballerina/modules/parser/parser.bal]: updated

File: ballerina/engine.bal (hunk 2 above)
Comment (MohamedSabthar) [reply ballerina/engine.bal]: updated

File: ballerina/websocket_utils.bal (hunk 4 above)
Comment (MohamedSabthar) [reply ballerina/websocket_utils.bal]: updated

File: ballerina/engine.bal (hunk 3 above)
Comment (MohamedSabthar) [reply ballerina/engine.bal]: updated

**Comment by ThisaruGuruge** [comment]
Check similar cases as mentioned in the comments. Let's resolve the conflicts as well.

File: ballerina/document_node_modifier_visitor.bal [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class DocumentNodeModifierVisitor {
+    *parser:Visitor;
+
+    private NodeModifierContext nodeModifierContext;
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/document_node_modifier_visitor.bal]: Updated

File: ballerina/document_node_modifier_visitor.bal [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    private parser:DocumentNode? document;
+    private map<parser:Node> modifiedNodes;
+
+    isolated function init(NodeModifierContext nodeModifierContext) {
+        self.nodeModifierContext = nodeModifierContext;
+        self.modifiedNodes = {};
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/document_node_modifier_visitor.bal]: updated

File: ballerina/document_node_modifier_visitor.bal [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+// under the License.
+
+import graphql.parser;
+
+class DocumentNodeModifierVisitor {
+    *parser:Visitor;
+
+    private NodeModifierContext nodeModifierContext;
+    private parser:DocumentNode? document;
+    private map<parser:Node> modifiedNodes;
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/document_node_modifier_visitor.bal]: updated

File: ballerina/modules/parser/utils.bal [hunk 8]
```diff
@@ -80,3 +80,7 @@ isolated function isValidFirstChar(string char) returns boolean = @java:Method {
 isolated function isValidChar(string char) returns boolean = @java:Method {
     'class: "io.ballerina.stdlib.graphql.runtime.parser.ParserUtils"
 } external;
+
+public isolated function getHashCode(object {} obj) returns string = @java:Method {
```
Comment (ThisaruGuruge) [inline ballerina/modules/parser/utils.bal:84]: Let's create an issue to track the progress of replacing this function with an ID field.

File: ballerina/websocket_service.bal [hunk 9]
```diff
@@ -91,6 +91,7 @@ isolated service class WsService {
                 if connectionId is () {
                     return self.handleIdNotPresentInPayload(caller);
                 }
+                SubscriptionHandler handler = new(connectionId);
```
Comment (ThisaruGuruge) [inline ballerina/websocket_service.bal:94]: Did we remove the support for non-protocol requests? If not, let's remove that with PR. Going forward, we should only allow the requests that adhere to the graphql-ws protocol.

File: ballerina/websocket_utils.bal [hunk 10]
```diff
@@ -19,36 +19,69 @@ import ballerina/websocket;
...
+            check handleStreamCreationError(customHeaders, caller, handler, sourceStream);
         }
-        closeStream(sourceStream);
-    } else {
-        check sendWebSocketResponse(caller, customHeaders, WS_ERROR, sourceStream, connectionId);
-        if !customHeaders.hasKey(WS_SUB_PROTOCOL) {
-            closeConnection(caller);
+    } on fail error err {
+        log:printError(err.message(), stackTrace = err.stackTrace());
+        if (sourceStream is stream<any, error?>) {
```
Comment (ThisaruGuruge) [inline ballerina/websocket_utils.bal:52]:
```suggestion
        if sourceStream is stream<any, error?> {
```

File: docs/spec/spec.md [hunk 11]
```diff
@@ -418,7 +418,7 @@ service on new graphql:Listener(9090) {
 }
 ```
 
->**Note:**: The current implementation does not support multiplexing between multiple subscription operations using a client.
+> **Note:**: A resource method with `subscribe` accessor must return a Ballerina `stream` type.
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md:421]:
```suggestion
>**Note:**: A resource method with a `subscribe` accessor must return a Ballerina `stream` type.
```

File: ballerina/websocket_service.bal (hunk 9 above)
Comment (MohamedSabthar) [reply ballerina/websocket_service.bal:94]: Support for non-protocol requests are not yet removed. We could address this change in a new PR. Related issue: https://github.com/ballerina-platform/ballerina-standard-library/issues/3608

File: ballerina/modules/parser/utils.bal (hunk 8 above)
Comment (MohamedSabthar) [reply ballerina/modules/parser/utils.bal:84]: Issue: https://github.com/ballerina-platform/ballerina-standard-library/issues/3881

File: ballerina/websocket_service.bal [hunk 12]
```diff
@@ -24,7 +24,7 @@ isolated service class WsService {
     private final Engine engine;
     private final readonly & __Schema schema;
     private final Context context;
-    private final map<()> activeConnections;
+    private map<SubscriptionHandler> activeConnections;
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/websocket_service.bal:27]: Yes, updated

File: ballerina/engine.bal [hunk 13]
```diff
@@ -48,23 +48,27 @@ isolated class Engine {
...
-        parser:DocumentNode|OutputObject result = self.parse(documentString);
+        ParseResult|OutputObject result = self.parse(documentString);
         if result is OutputObject {
             return result;
         }
-        parser:DocumentNode document = <parser:DocumentNode>result;
-        OutputObject? validationResult = self.validateDocument(document, variables);
+        parser:DocumentNode document = result.document;
+        ErrorDetail[] validationErrors = result.validationErrors;
+        OutputObject|parser:DocumentNode validationResult = self.validateDocument(document, variables, validationErrors);
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/engine.bal:57]: updated

File: ballerina/engine.bal [hunk 14]
```diff
@@ -48,23 +48,27 @@ isolated class Engine {
...
             return self.getOperation(document, operationName);
         }
     }
 
     isolated function getResult(parser:OperationNode operationNode, Context context, any|error result = ())
     returns OutputObject {
-        DefaultDirectiveProcessorVisitor defaultDirectiveProcessor = new (self.schema);
-        DuplicateFieldRemoverVisitor duplicateFieldRemover = new;
+        final map<()> removedNodes = {};
+        final map<parser:SelectionNode> modifiedSelections = {};
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/engine.bal:69]: Updatd

File: docs/spec/spec.md [hunk 15]
```diff
@@ -418,7 +418,7 @@ service on new graphql:Listener(9090) {
 }
 ```
 
->**Note:**: The current implementation does not support multiplexing between multiple subscription operations using a client.
+>**Note:**: A resource method with a `subscribe` accessor must return a Ballerina `stream` type.
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply docs/spec/spec.md:421]: removed

File: ballerina/engine.bal [hunk 16]
```diff
@@ -48,23 +48,27 @@ isolated class Engine {
...
-        parser:DocumentNode document = <parser:DocumentNode>result;
-        OutputObject? validationResult = self.validateDocument(document, variables);
+        parser:DocumentNode document = result.document;
+        ErrorDetail[] validationErrors = result.validationErrors;
+        OutputObject|parser:DocumentNode validationResult = self.validateDocument(document, variables, validationErrors);
         if validationResult is OutputObject {
             return validationResult;
         } else {
+            document = validationResult;
             return self.getOperation(document, operationName);
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply ballerina/engine.bal:62]: removed

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**